File: examples/pendulum/pendulum.py
import os

import numpy as np
from numba import njit

import madupite as md

# Constants and model parameters
NUM_X = 201  # Number of discretized positions (angles)
NUM_XD = 201  # Number of discretized angular velocities
NUM_A = 9  # Number of discretized actions/torque

MAX_X = 2 * np.pi  # Maximum position (angle)
MAX_XD = 10.0  # Maximum angular velocity
MAX_A = 3.0  # Maximum torque
R = 1  # Penalty for control effort
Q = 2  # Penalty for state deviation
DT = 0.01  # Time step
G = 9.81  # Gravitational acceleration
L = 1.0  # Pendulum length
M = 1.0  # Pendulum mass
NUM_STATES = NUM_X * NUM_XD  # Total number of states
NUM_ACTIONS = NUM_A  # Total number of actions


# Helper functions that map a 2D state onto a 1D index
@njit
def x2s(x, xd):
    return int(x * NUM_XD + xd)


@njit
def s2x(s):
    return s // NUM_XD, s % NUM_XD


@njit
def numba_clip(value, min_value, max_value):
    """Clip value within min_value and max_value."""
    return min(max(value, min_value), max_value)


@njit
def interpolate(x, y, grid_x, grid_y):
    """Interpolate between grid points."""
    x_i = np.searchsorted(grid_x, x, side="right")
    y_i = np.searchsorted(grid_y, y, side="right")
    x_i, y_i = numba_clip(x_i, 1, len(grid_x) - 1), numba_clip(y_i, 1, len(grid_y) - 1)
    xl_v, xr_v = grid_x[x_i - 1], grid_x[x_i]
    yl_v, yr_v = grid_y[y_i - 1], grid_y[y_i]
    wx1, wy1 = (x - xl_v) / (xr_v - xl_v), (y - yl_v) / (yr_v - yl_v)
    wx0, wy0 = 1 - wx1, 1 - wy1
    indices = [(x_i - 1, y_i - 1), (x_i, y_i - 1), (x_i - 1, y_i), (x_i, y_i)]
    weights = [wx0 * wy0, wx1 * wy0, wx0 * wy1, wx1 * wy1]
    return indices, weights


@njit
def step(s, a):
    """Calculate the next state for a given state-action pair."""
    x_t_i, xd_t_i = s2x(s)
    x_t_v = x_vals[x_t_i]
    xd_t_v = xd_vals[xd_t_i]
    a_t_v = a_vals[a]

    # Compute next state based on dynamics
    x_tpp_v = (x_t_v + xd_t_v * DT) % MAX_X  # Periodic boundary condition
    xdd_t_v = -G / L * np.sin(x_t_v) + a_t_v / (M * L**2)
    xd_tpp_v = numba_clip(xd_t_v + xdd_t_v * DT, -MAX_XD, MAX_XD)

    return x_tpp_v, xd_tpp_v


def stage_cost(s, a):
    """Compute stage cost for a given state-action pair."""
    x_i, xd_i = s2x(s)
    return Q * ((x_vals[x_i] - np.pi) ** 2 + xd_vals[xd_i] ** 2) + R * a_vals[a] ** 2


def transition_probability(s, a):
    """Compute transition probabilities for a given state-action pair."""
    x_tpp_v, xd_tpp_v = step(s, a)
    indices, weights = interpolate(x_tpp_v, xd_tpp_v, x_vals, xd_vals)
    s_indices = [x2s(x_i, xd_i) for x_i, xd_i in indices]
    return weights, s_indices


def main():
    # Initialize grid values
    global x_vals, xd_vals, a_vals
    x_vals = np.linspace(0, MAX_X, NUM_X)  # Starting from 0
    xd_vals = np.linspace(-MAX_XD, MAX_XD, NUM_XD)
    a_vals = np.linspace(-MAX_A, MAX_A, NUM_A)

    # Initialize Madupite instance
    madupite_instance = md.initialize_madupite()

    # Create MDP instance
    mdp = md.MDP(madupite_instance)

    # Set MDP options
    mdp.setOption("-mode", "MINCOST")
    mdp.setOption("-max_iter_pi", "100")
    mdp.setOption("-max_iter_ksp", "20000")
    mdp.setOption("-alpha", "1e-5")
    mdp.setOption("-atol_pi", "1e-7")
    mdp.setOption("-ksp_type", "tfqmr")
    mdp.setOption("-discount_factor", "0.99")
    mdp.setOption("-file_stats", "data/pend_stats.json")
    mdp.setOption("-file_policy", "data/pend_policy.out")
    mdp.setOption("-file_cost", "data/pend_cost.out")
    mdp.setOption("-export_optimal_transition_probabilities", "data/pend_Ppi.out")
    mdp.setOption("-export_optimal_stage_costs", "data/pend_gpi.out")

    # Create the stage cost matrix using the stage cost function
    filename_P = f"data/pend_P_{NUM_STATES}_{NUM_ACTIONS}.bin"
    filename_g = f"data/pend_g_{NUM_STATES}_{NUM_ACTIONS}.bin"
    if os.path.exists(filename_P):
        P_mat = md.Matrix.fromFile(
            comm=md.getCommWorld(),
            name="P",
            filename=filename_P,
            category=md.MatrixCategory.Dynamics,
            type=md.MatrixType.Sparse,
        )
    else:
        P_mat = md.createTransitionProbabilityTensor(
            name="P",
            numStates=NUM_STATES,
            numActions=NUM_ACTIONS,
            func=transition_probability,
        )
        P_mat.writeToFile(filename_P, md.MatrixType.Sparse, True)

    if os.path.exists(filename_g):
        g_mat = md.Matrix.fromFile(
            comm=md.getCommWorld(),
            name="g",
            filename=filename_g,
            category=md.MatrixCategory.Cost,
            type=md.MatrixType.Dense,
        )
    else:
        g_mat = md.createStageCostMatrix(
            name="g", numStates=NUM_STATES, numActions=NUM_ACTIONS, func=stage_cost
        )
        g_mat.writeToFile(filename_g, md.MatrixType.Dense, True)

    # Set matrices in the MDP
    mdp.setStageCostMatrix(g_mat)
    mdp.setTransitionProbabilityTensor(P_mat)

    if md.mpi_rank_size()[0] == 0:
        print("Solving...", flush=True)

    # Solve the MDP
    mdp.solve()

    del mdp


if __name__ == "__main__":
    main()

File: examples/StrongScaling/pendulum.cpp
#include <algorithm>
#include <cmath>
#include <functional>
#include <numbers>
#include <string>
#include <vector>

#include "MDP.h"

class Pendulum {
public:
    Pendulum(int num_x = 401, int num_xd = 401, int num_a = 201, double max_x = 2 * std::numbers::pi, double max_xd = 10.0, double max_a = 3.0,
        double r = 1.0, double q = 2.0, double dt = 0.01, double g = 9.81, double l = 1.0, double m = 1.0)
        : NUM_X(num_x)
        , NUM_XD(num_xd)
        , NUM_A(num_a)
        , MAX_X(max_x)
        , MAX_XD(max_xd)
        , MAX_A(max_a)
        , R(r)
        , Q(q)
        , DT(dt)
        , G(g)
        , L(l)
        , M(m)
        , NUM_STATES(NUM_X * NUM_XD)
        , NUM_ACTIONS(NUM_A)
    {

        init_vals();
    }

    std::function<std::pair<std::vector<double>, std::vector<int>>(int, int)> P_func() const
    {
        return [this](int s, int a) { return this->P(s, a); };
    }

    std::function<double(int, int)> g_func() const
    {
        return [this](int s, int a) { return this->g(s, a); };
    }

    const std::vector<double>& get_x_vals() const { return x_vals; }
    const std::vector<double>& get_xd_vals() const { return xd_vals; }
    const std::vector<double>& get_a_vals() const { return a_vals; }
    const int                  get_num_states() const { return NUM_STATES; }
    const int                  get_num_actions() const { return NUM_ACTIONS; }

private:
    int x2s(int x, int xd) const { return x * NUM_XD + xd; }

    std::pair<int, int> s2x(int s) const { return { s / NUM_XD, s % NUM_XD }; }

    std::pair<std::vector<std::pair<int, int>>, std::vector<double>> interpolate(
        double x, double y, const std::vector<double>& grid_x, const std::vector<double>& grid_y) const
    {
        int x_i = std::upper_bound(grid_x.begin(), grid_x.end(), x) - grid_x.begin();
        int y_i = std::upper_bound(grid_y.begin(), grid_y.end(), y) - grid_y.begin();

        x_i = std::clamp(x_i, 1, static_cast<int>(grid_x.size()) - 1);
        y_i = std::clamp(y_i, 1, static_cast<int>(grid_y.size()) - 1);

        double xl_v = grid_x[x_i - 1], xr_v = grid_x[x_i];
        double yl_v = grid_y[y_i - 1], yr_v = grid_y[y_i];

        double wx1 = (x - xl_v) / (xr_v - xl_v);
        double wy1 = (y - yl_v) / (yr_v - yl_v);
        double wx0 = 1 - wx1, wy0 = 1 - wy1;

        std::vector<std::pair<int, int>> indices = { { x_i - 1, y_i - 1 }, { x_i, y_i - 1 }, { x_i - 1, y_i }, { x_i, y_i } };
        std::vector<double>              weights = { wx0 * wy0, wx1 * wy0, wx0 * wy1, wx1 * wy1 };

        return { indices, weights };
    }

    std::pair<double, double> step(int x_t_i, int xd_t_i, int a) const
    {
        double x_t_v  = x_vals[x_t_i];
        double xd_t_v = xd_vals[xd_t_i];
        double a_t_v  = a_vals[a];

        double x_tpp_v  = circular_mod(x_t_v + xd_t_v * DT, MAX_X); // periodic boundary!
        double xdd_t_v  = -G / L * std::sin(x_t_v) + a_t_v / (M * L * L);
        double xd_tpp_v = std::clamp(xd_t_v + xdd_t_v * DT, -MAX_XD, MAX_XD);

        return { x_tpp_v, xd_tpp_v };
    }

    double stage_costs(int x_i, int xd_i, int a) const
    {
        return Q * (std::pow(x_vals[x_i] - std::numbers::pi, 2) + std::pow(xd_vals[xd_i], 2)) + R * std::pow(a_vals[a], 2);
    }

    std::pair<std::vector<double>, std::vector<int>> P(int s, int a) const
    {
        auto [x_t_i, xd_t_i]     = s2x(s);
        auto [x_tpp_v, xd_tpp_v] = step(x_t_i, xd_t_i, a);

        auto [indices, weights] = interpolate(x_tpp_v, xd_tpp_v, x_vals, xd_vals);

        std::vector<int> next_states;
        for (const auto& [x_i, xd_i] : indices) {
            next_states.push_back(x2s(x_i, xd_i));
        }

        return { weights, next_states };
    }

    double g(int s, int a) const
    {
        auto [x_t_i, xd_t_i] = s2x(s);
        return stage_costs(x_t_i, xd_t_i, a);
    }

    double circular_mod(double a, double b) const
    {
        double result = std::fmod(a, b);
        return result >= 0 ? result : result + b;
    }

    void init_vals()
    {
        x_vals.resize(NUM_X);
        xd_vals.resize(NUM_XD);
        a_vals.resize(NUM_A);

        for (int i = 0; i < NUM_X; ++i) {
            x_vals[i] = i * MAX_X / (NUM_X - 1);
        }
        for (int i = 0; i < NUM_XD; ++i) {
            xd_vals[i] = -MAX_XD + i * 2 * MAX_XD / (NUM_XD - 1);
        }
        for (int i = 0; i < NUM_A; ++i) {
            a_vals[i] = -MAX_A + i * 2 * MAX_A / (NUM_A - 1);
        }
    }

    const int    NUM_X, NUM_XD, NUM_A;
    const double MAX_X, MAX_XD, MAX_A;
    const double R, Q, DT, G, L, M;
    const int    NUM_STATES, NUM_ACTIONS;

    std::vector<double> x_vals, xd_vals, a_vals;
};

int main(int argc, char** argv)
{
    // Initialize MPI, PETSc and Madupite, passing command line arguments.
    auto madupite = Madupite::initialize(&argc, &argv);

    Pendulum pendulum(401, 401, 201);
    MDP      mdp(madupite);

    mdp.setOption("-mode", "MINCOST");
    mdp.setOption("-max_iter_pi", "200");
    mdp.setOption("-max_iter_ksp", "2000");
    mdp.setOption("-alpha", "1e-2");
    mdp.setOption("-ksp_type", "tfqmr");
    mdp.setOption("-atol_pi", "1e-9");
    mdp.setOption("-num_states", std::to_string(pendulum.get_num_states()).c_str());
    mdp.setOption("-num_actions", std::to_string(pendulum.get_num_actions()).c_str());
    mdp.setOption("-discount_factor", "0.999");

    // mdp.setOption("-file_stats", "pend_stats.json");
    // mdp.setOption("-file_policy", "pend_policy.out");
    // mdp.setOption("-file_cost", "pend_cost.out");

    mdp.setSourceTransitionProbabilityTensor(pendulum.P_func(), 4, {}, 4, {});
    mdp.setSourceStageCostMatrix(pendulum.g_func());

    mdp.solve();
}
